tests/phpunit/tests/html-api/wpHtmlProcessor.php

```diff
@@ -882,4 +882,180 @@ public function test_ensure_form_tag_closer_token_is_reachable() {
+					8  => '/*[1][self::HTML]/*[2][self::BODY]/*[5][self::UL]/*[1][self::LI]',
+					9  => '/*[1][self::HTML]/*[2][self::BODY]/*[5][self::UL]/*[2][self::LI]',
+					10 => '/*[1][self::HTML]/*[2][self::BODY]/*[5][self::UL]/*[3][self::LI]',
+				),
+			),
+
+			'extreme_nested_formatting' => array(
+				'html'                  => '
+					<html>
+						<body>
+							<p>
+								<strong><em><strike><i><b><u>FORMAT</u></b></i></strike></em></strong>
+							</p>
+						</body>
+					</html>
+				',
+				'expected_token_counts' => array(
+					'+HTML'   => 1,
+					'+HEAD'   => 1,
+					'-HEAD'   => 1,
+					'+BODY'   => 1,
+					'#text'   => 7,
+					'+P'      => 1,
+					'+STRONG' => 1,
+					'+EM'     => 1,
+					'+STRIKE' => 1,
+					'+I'      => 1,
+					'+B'      => 1,
+					'+U'      => 1,
+					'-U'      => 1,
+					'-B'      => 1,
+					'-I'      => 1,
+					'-STRIKE' => 1,
+					'-EM'     => 1,
+					'-STRONG' => 1,
+					'-P'      => 1,
+					'-BODY'   => 1,
+					'-HTML'   => 1,
+					''        => 1,
+				),
+				'expected_xpaths'       => array(
+					0 => '/*[1][self::HTML]',
+					1 => '/*[1][self::HTML]/*[1][self::HEAD]',
+					2 => '/*[1][self::HTML]/*[2][self::BODY]',
+					3 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]',
+					4 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]',
+					5 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]/*[1][self::EM]',
+					6 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]/*[1][self::EM]/*[1][self::STRIKE]',
+					7 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]/*[1][self::EM]/*[1][self::STRIKE]/*[1][self::I]',
+					8 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]/*[1][self::EM]/*[1][self::STRIKE]/*[1][self::I]/*[1][self::B]',
+					9 => '/*[1][self::HTML]/*[2][self::BODY]/*[1][self::P]/*[1][self::STRONG]/*[1][self::EM]/*[1][self::STRIKE]/*[1][self::I]/*[1][self::B]/*[1][self::U]',
+				),
+			),
+		);
+	}
+
+	/**
+	 * Ensures that subclasses to WP_HTML_Processor can do bookkeeping by extending the next_token() method.
+	 *
+	 * @ticket ?
+	 * @dataProvider data_html_processor_with_extended_next_token
+	 */
+	public function test_ensure_next_token_method_extensibility( $html, $expected_token_counts, $expected_xpaths ) {
+		require_once DIR_TESTDATA . '/html-api/html-xpath-generating-processor.php';
+
+		$processor     = HTML_XPath_Generating_Processor::create_full_parser( $html );
+		$actual_xpaths = array();
+		while ( $processor->next_tag() ) {
+			if ( ! $processor->is_tag_closer() ) {
+				$processor->set_attribute( 'xpath', $processor->get_xpath() );
+				$actual_xpaths[] = $processor->get_xpath();
+			}
+		}
+
+		$this->assertEquals( $expected_token_counts, $processor->token_seen_count, 'Snapshot: ' . var_export( $processor->token_seen_count, true ) );
+		$this->assertEquals( $expected_xpaths, $actual_xpaths, 'Snapshot: ' . var_export( $actual_xpaths, true ) );
+	}
```

Comment:
this is an oddly-specific unit test suite. I'm trying to think through how difficult it could be to debug a test failure given that we've imported the full mechanism from the image optimization plugin - it's kind of a misdirection here, hinting that someone might need to understand XPath and what's going on, and why the XPath queries are the specific shape they are.

is there a way we could think of a more minimal test to cover this behavior? perhaps a subclass with a reimplementation of `next_token()` that does something simple, like report a count of tokens?

then, hopefully we could test on a short string like `<p>text` and demonstrate the failure before the patch and success after it.

Reply:
See https://github.com/WordPress/wordpress-develop/pull/7730